File: practica6/Movimiento.cpp
#include "Movimiento.h"
#include "Pokemon.h"
#include "global.h"
#include "Calculadores.h"
#include <iostream>

Movimiento::Movimiento(std::string n, std::shared_ptr<TipoElemento> t, int p,
                       const IMostrador* m,
                       std::shared_ptr<ICalculadorDano> calcDano,
                       std::shared_ptr<ICalculadorMultiplicador> calcMult)
    : nombre(n), tipo(t), potencia(p), mostrador(m),
      calculadorDano(calcDano ? calcDano : std::make_shared<CalculadorDanoBasico>()),
      calculadorMultiplicador(calcMult ? calcMult : std::make_shared<CalculadorMultiplicadorFuerza>()) {}

std::string Movimiento::getNombre() const { return nombre; }
std::shared_ptr<TipoElemento> Movimiento::getTipo() const { return tipo; }
int Movimiento::getPotencia() const { return potencia; }

// SRP: Método con responsabilidad única - calcular multiplicador
double Movimiento::obtenerMultiplicadorTipo(const Pokemon& objetivo) const {
    double fuerzaAtaque = tipo->getFuerza();
    
    double fuerzaDefensaTotal = 0.0;
    for (const auto& tipoDefensor : objetivo.getTipos()) {
        fuerzaDefensaTotal += tipoDefensor->getFuerza();
    }
    double fuerzaDefensaPromedio = fuerzaDefensaTotal / objetivo.getTipos().size();
    
    // DIP: Usa abstracción para calcular
    return calculadorMultiplicador->calcularMultiplicador(fuerzaAtaque, fuerzaDefensaPromedio);
}

// SRP: Método con responsabilidad única - mostrar mensajes
void Movimiento::mostrarMensajeEfectividad(double multiplicador) const {
    if (!mostrador) return;
    
    if (multiplicador >= 1.5) {
        mostrador->mostrarMensaje("¡El tipo es muy fuerte contra el oponente!", YELLOW);
    } else if (multiplicador >= 1.2) {
        mostrador->mostrarMensaje("El tipo tiene ventaja.", GREEN);
    } else if (multiplicador <= 0.7) {
        mostrador->mostrarMensaje("El tipo es debil contra el oponente...", CYAN);
    } else if (multiplicador <= 0.9) {
        mostrador->mostrarMensaje("El tipo tiene desventaja.", CYAN);
    }
}

void Movimiento::aplicarEfecto(Pokemon& atacante, Pokemon& objetivo) const {
    if (potencia > 0) {
        // SRP: Cada cálculo delegado a su clase responsable
        double multiplicador = obtenerMultiplicadorTipo(objetivo);
        
        // DIP: Usa abstracción para calcular daño
        int dmg = calculadorDano->calcularDano(
            atacante.getAtaque(),
            potencia,
            objetivo.getDefensa(),
            multiplicador
        );

        // DIP: Usa abstracción para mostrar mensajes
        if (mostrador) {
            mostrador->mostrarMensajeCombate(
                atacante.getNombre() + " usa " + nombre + 
                " (" + tipo->getNombre() + ", Fuerza: " + std::to_string(static_cast<int>(tipo->getFuerza())) + 
                ") y causa " + std::to_string(dmg) + 
                " de dano a " + objetivo.getNombre() + "."
            );
            
            mostrarMensajeEfectividad(multiplicador);
        }

        objetivo.recibirDano(dmg);
    } else {
        if (mostrador) {
            mostrador->mostrarMensaje(atacante.getNombre() + " usa " + nombre + ".");
        }
    }
}

// OCP: Extensión de funcionalidad
MovimientoEstado::MovimientoEstado(std::string n, std::shared_ptr<TipoElemento> t, int p,
                                   const IMostrador* m,
                                   std::shared_ptr<ICalculadorDano> calcDano,
                                   std::shared_ptr<ICalculadorMultiplicador> calcMult)
    : Movimiento(n, t, p, m, calcDano, calcMult) {}

void MovimientoEstado::aplicarEfecto(Pokemon& atacante, Pokemon& objetivo) const {
    Movimiento::aplicarEfecto(atacante, objetivo);
    if (mostrador) {
        mostrador->mostrarMensaje(objetivo.getNombre() + " ha sido envenenado!", MAGENTA);
    }
}

// OCP: Extensión de funcionalidad
MovimientoCuracion::MovimientoCuracion(std::string n, std::shared_ptr<TipoElemento> t, int p, int curacion,
                                       const IMostrador* m)
    : Movimiento(n, t, p, m), cantidadCuracion(curacion) {}

void MovimientoCuracion::aplicarEfecto(Pokemon& atacante, Pokemon& objetivo) const {
    if (mostrador) {
        mostrador->mostrarMensaje(
            atacante.getNombre() + " usa " + nombre +
            " y se cura " + std::to_string(cantidadCuracion) + " puntos de vida!", GREEN
        );
    }
    atacante.curar(cantidadCuracion);
}

File: pokemon/movimientos/Movimiento.java
package pokemon.movimientos;

import pokemon.Pokemon;
import pokemon.tipos.TipoElemento;
import pokemon.interfaces.IMostrador;
import pokemon.interfaces.ICalculadorDano;
import pokemon.interfaces.ICalculadorMultiplicador;
import pokemon.calculadores.CalculadorDanoBasico;
import pokemon.calculadores.CalculadorMultiplicadorFuerza;
import static pokemon.Global.*;

public class Movimiento {
    protected String nombre;
    protected TipoElemento tipo;
    protected int potencia;
    
    // DIP: Dependencias son abstracciones
    protected IMostrador mostrador;
    protected ICalculadorDano calculadorDano;
    protected ICalculadorMultiplicador calculadorMultiplicador;
    
    public Movimiento(String nombre, TipoElemento tipo, int potencia,
                     IMostrador mostrador,
                     ICalculadorDano calcDano,
                     ICalculadorMultiplicador calcMult) {
        this.nombre = nombre;
        this.tipo = tipo;
        this.potencia = potencia;
        this.mostrador = mostrador;
        this.calculadorDano = calcDano != null ? calcDano : new CalculadorDanoBasico();
        this.calculadorMultiplicador = calcMult != null ? calcMult : new CalculadorMultiplicadorFuerza();
    }
    
    public String getNombre() {
        return nombre;
    }
    
    public TipoElemento getTipo() {
        return tipo;
    }
    
    public int getPotencia() {
        return potencia;
    }
    
    // SRP: Método con responsabilidad única - calcular multiplicador
     
    protected double obtenerMultiplicadorTipo(Pokemon objetivo) {
        double fuerzaAtaque = tipo.getFuerza();
        
        double fuerzaDefensaTotal = 0.0;
        for (TipoElemento tipoDefensor : objetivo.getTipos()) {
            fuerzaDefensaTotal += tipoDefensor.getFuerza();
        }
        double fuerzaDefensaPromedio = fuerzaDefensaTotal / objetivo.getTipos().size();
        
        // DIP: Usa abstracción para calcular
        return calculadorMultiplicador.calcularMultiplicador(fuerzaAtaque, fuerzaDefensaPromedio);
    }
    
    //SRP: Método con responsabilidad única - mostrar mensajes
    protected void mostrarMensajeEfectividad(double multiplicador) {
        if (mostrador == null) return;
        
        if (multiplicador >= 1.5) {
            mostrador.mostrarMensaje("¡El tipo es muy fuerte contra el oponente!", YELLOW);
        } else if (multiplicador >= 1.2) {
            mostrador.mostrarMensaje("El tipo tiene ventaja.", GREEN);
        } else if (multiplicador <= 0.7) {
            mostrador.mostrarMensaje("El tipo es débil contra el oponente...", CYAN);
        } else if (multiplicador <= 0.9) {
            mostrador.mostrarMensaje("El tipo tiene desventaja.", CYAN);
        }
    }
    
    //OCP: Método virtual para extensión
    
    public void aplicarEfecto(Pokemon atacante, Pokemon objetivo) {
        if (potencia > 0) {
            // SRP: Cada cálculo delegado a su clase responsable
            double multiplicador = obtenerMultiplicadorTipo(objetivo);
            
            // DIP: Usa abstracción para calcular daño
            int dmg = calculadorDano.calcularDano(
                atacante.getAtaque(),
                potencia,
                objetivo.getDefensa(),
                multiplicador
            );
            
            // DIP: Usa abstracción para mostrar mensajes
            if (mostrador != null) {
                mostrador.mostrarMensajeCombate(
                    atacante.getNombre() + " usa " + nombre +
                    " (" + tipo.getNombre() + ", Fuerza: " + (int) tipo.getFuerza() +
                    ") y causa " + dmg + " de daño a " + objetivo.getNombre() + "."
                );
                
                mostrarMensajeEfectividad(multiplicador);
            }
            
            objetivo.recibirDano(dmg);
        } else {
            if (mostrador != null) {
                mostrador.mostrarMensaje(atacante.getNombre() + " usa " + nombre + ".",RESET );
            }
        }
    }
}